libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/RouteRefreshController.kt

```diff
@@ -1,79 +1,141 @@
     }
 
-    private val routerRefreshTimer = MapboxTimer(threadController).apply {
-        restartAfterMillis = routeRefreshOptions.intervalMillis
+    suspend fun refresh(routes: List<NavigationRoute>): List<NavigationRoute> {
+        return if (routes.firstOrNull()?.routeOptions?.enableRefresh() == true) {
+            val result = routes.toMutableList()
+            val routeToRefresh = result.first()
+            result[0] = tryRefreshingUntilRouteChanges(routeToRefresh)
+            result
+        } else waitForever("refresh isn't enabled")
     }
 
-    private var currentRequestId: Long? = null
-
-    /**
-     * If route refresh is enabled, attach a refresh poller.
-     * Cancel old timers, and cancel pending requests.
-     */
-    fun restart(route: NavigationRoute, onRoutesRefreshed: (List<NavigationRoute>) -> Unit) {
-        stop()
-        if (route.routeOptions.enableRefresh() == true) {
-            routerRefreshTimer.startTimer {
-                refreshRoute(route, onRoutesRefreshed)
+    private suspend fun tryRefreshingUntilRouteChanges(
+        initialRoute: NavigationRoute
+    ): NavigationRoute {
+        while (true) {
+            val refreshed = refreshRoute(initialRoute)
+            if (refreshed != initialRoute) {
+                return refreshed
             }
         }
     }
 
-    /**
-     * Cancel old timers, and cancel pending requests.
-     */
-    fun stop() {
-        routerRefreshTimer.stopJobs()
-        currentRequestId?.let {
-            directionsSession.cancelRouteRefreshRequest(it)
-            currentRequestId = null
+    private suspend fun refreshRoute(
+        route: NavigationRoute
+    ): NavigationRoute = coroutineScope {
+        val routeLegs = route.directionsRoute.legs()
+        require(routeLegs != null) { "Can't refresh route without legs" }
+
+        var timeUntilNextAttempt = async { delay(routeRefreshOptions.intervalMillis) }
+        for (attempts in 0..2) {
+            timeUntilNextAttempt.await()
+            timeUntilNextAttempt = async { delay(routeRefreshOptions.intervalMillis) }
+            val refreshedRoute = withTimeoutOrNull(routeRefreshOptions.intervalMillis) {
+                refreshRouteOrNull(route)
+            }
+            if (refreshedRoute != null) {
+                timeUntilNextAttempt.cancel()
+                return@coroutineScope refreshedRoute
+            }
         }
+        timeUntilNextAttempt.cancel()
+        removeTrafficAnnotations(route, routeLegs)
     }
 
-    private fun refreshRoute(
+    private fun removeTrafficAnnotations(
         route: NavigationRoute,
-        onRoutesRefreshed: (List<NavigationRoute>) -> Unit
-    ) {
+        routeLegs: List<RouteLeg>
+    ): NavigationRoute {
+        val currentLegIndex = currentLegIndexProvider()
+        return route.refreshRoute(
+            0,
+            routeLegs.mapIndexed { index, leg ->
+                if (index == currentLegIndex) {
```

Comment:
all annotations must be invalidated, not only for the current leg

Reply:
Invalidation will be inconsistent with the refresh feature then, won't it? It also raises a question of why not to invalidate annotations on other legs even if route refresh for current is successful 🤔 
What do you think?

Reply:
not sure, why you think it's inconsistent 🤔 Route refresh is working for the whole route, starting with leg `x`, so if the SDK cannot refresh the route, all annotations should be cleaned up. This is applicable for a single route (whenever primary or alternative)

Reply:
done